java/spark-1.5.2/rutils.scala

```diff
@@ -0,0 +1,161 @@
+
+    val num_rows = readLength(dis)
+
+    (0L until num_rows).map(
+      r => {
+        val elem_dtype = readDataType(dis)
+
+        if (NILVALUE_SXP == elem_dtype) {
+          null
+        } else {
+          val num_elems = readLength(dis)
+
+          if (1 != num_elems)
+            throw new IllegalArgumentException(
+              s"Unexpected number of elements: $num_elems"
+            )
+
+          elem_dtype match {
+            case LGLSXP => {
+              val v = Serializer.readInt(dis)
+              if (NA_INTEGER == v) null else v != 0
+            }
+            case INTSXP => {
+              val v = Serializer.readInt(dis)
+              if (NA_INTEGER == v) null else v
+            }
+            case REALSXP => {
+              val bytes = new Array[Byte](8)
+              dis.readFully(bytes)
+              val hw: Long = extractInt(bytes, 0) & 0x00000000FFFFFFFFL
+              val lw: Long = extractInt(bytes, 4) & 0x00000000FFFFFFFFL
+              if (hw == 0x7FF00000L && lw == 1954L)
+                null
+              else
+                java.lang.Double.longBitsToDouble((hw << 32) | lw)
+            }
+            case STRSXP => {
+              val ctype = readDataType(dis)
+              if (CHARSXP != ctype)
+                logger.logWarning(
+                  s"Unexpected character type $ctype found in string expression"
+                )
+
+              /* these are currently limited to 2^31 -1 bytes */
+              val strlen = Serializer.readInt(dis)
+              if (-1L == strlen) {
+                null
+              } else {
+                val bytes = new Array[Byte](strlen.intValue)
+                dis.readFully(bytes)
+
+                new String(bytes, StandardCharsets.UTF_8)
+              }
+            }
+            case _ =>
+              throw new IllegalArgumentException(s"Invalid type $elem_dtype")
+          }
+        }
+      }.asInstanceOf[Any]
+    ).toArray
+  }
+
+  private[this] def unserializeIntColumn(
+    dis: DataInputStream
+  ) : Array[Int] = {
+    Array.fill[Int](5)(0)
+  }
+
+  private[this] def unserializeStringColumn(
+    dis: DataInputStream
+  ) : Array[String] = {
+    Array.fill[String](5)("")
+  }
+
+  private[this] def readDataType(
+    dis: DataInputStream
+  ): Int = {
+    Serializer.readInt(dis) & 0xFF
+  }
+
+  private[this] def readLength(
+    dis: DataInputStream
+  ): Long = {
+    val x = Serializer.readInt(dis)
+    val len = (
+      if (-1 == x) {
+        val upper: Long = Serializer.readInt(dis)
+        val lower: Long = Serializer.readInt(dis)
+        (upper << 32) + lower
+      } else {
+        x
+      }
+    )
+
+    if (len < 0)
+      throw new IllegalArgumentException(
+        s"Negative serialized vector length: $len"
+      )
+
+    len
+  }
+
+  private[this] def extractInt(bytes: Array[Byte], offset: Int): Int = {
+    (0 until 4).map(
+      i => (bytes(offset + i).intValue & 0xFF) << (24 - i * 8)
```

Comment:
`& 0xFF` due to having to mask off `1`s in high-order bits from sign extension -- it's super awkward when everything needs to be signed in Scala / Java